src/openms/source/ANALYSIS/ID/SiriusAdapterAlgorithm.cpp

```diff
@@ -0,0 +1,361 @@
+      }
+
+      // normalize file path
+      QString exe = executable.toQString();
+      QFileInfo file_info(exe);
+      exe = file_info.canonicalFilePath();
+  
+      LOG_WARN << "Executable is: " + String(exe) << std::endl;
+      const String path_to_executable = File::path(exe);
+      executable_workdir = std::make_pair(exe.toStdString(), path_to_executable);
+      
+      return executable_workdir;
+    }
+
+    SiriusAdapterAlgorithm::SiriusTmpStruct SiriusAdapterAlgorithm::constructSiriusTmpStruct()
+    {
+      SiriusTmpStruct tmp_struct;
+      QString base_dir = File::getTempDirectory().toQString();
+      tmp_struct.tmp_dir = String(QDir(base_dir).filePath(File::getUniqueName().toQString()));
+      tmp_struct.tmp_ms_file = QDir(base_dir).filePath((File::getUniqueName() + ".ms").toQString());
+      tmp_struct.tmp_out_dir = QDir(tmp_struct.tmp_dir.toQString()).filePath("sirius_out");
+
+      return tmp_struct;
+    } 
+    
+    void SiriusAdapterAlgorithm::preprocessingSirius(const String& featureinfo,
+                                                     MSExperiment& spectra,
+                                                     std::vector<FeatureMap>& v_fp,
+                                                     KDTreeFeatureMaps& fp_map_kd,
+                                                     const SiriusAdapterAlgorithm& sirius_algo,
+                                                     FeatureMapping::FeatureToMs2Indices& feature_mapping)
+    {
+      // if fileparameter is given and should be not empty
+      if (!featureinfo.empty())
+      {
+        if (File::exists(featureinfo) && !File::empty(featureinfo))
+        {
+          // read featureXML          
+          FeatureXMLFile fxml;
+          FeatureMap feature_map;
+          fxml.load(featureinfo, feature_map);
+          
+          bool feature_only = (sirius_algo.feature_only_ == "true") ? true : false;
+          unsigned int num_masstrace_filter = sirius_algo.filter_by_num_masstraces_;
+          double precursor_mz_tol = sirius_algo.precursor_mz_tolerance_;
+          double precursor_rt_tol = sirius_algo.precursor_rt_tolerance_;
+          bool ppm_prec = (sirius_algo.precursor_mz_tolerance_unit_ == "true") ? true : false; 
+          
+          if (num_masstrace_filter != 1 && !feature_only)
+          {
+            num_masstrace_filter = 1;
+            LOG_WARN << "Parameter: filter_by_num_masstraces, was set to 1 to retain the adduct information for all MS2 spectra, if available. Please use the masstrace filter in combination with feature_only." << std::endl;
+          }
+
+          // filter feature by number of masstraces
+          auto map_it = remove_if(feature_map.begin(), feature_map.end(),
+                                  [&num_masstrace_filter](const Feature &feat) -> bool
+                                  {
+                                    unsigned int n_masstraces = feat.getMetaValue("num_of_masstraces");
```

Comment:
this will be very slow, better store `feat.getMetaValue("num_of_masstraces");` in a cache variable